src/github.com/couchbase/sync_gateway/base/bucket_gocb.go

```diff
@@ -93,91 +97,416 @@ func (bucket CouchbaseBucketGoCB) GetName() string {
-		for _, key := range batchKeys {
-			var value []byte
-			item := &gocb.GetOp{Key: key, Value: &value}
-			items = append(items, item)
 		}
-		err := bucket.Do(items)
-		if err != nil {
-			Warn("Bulk get error for batch size %d:%v", len(items), err)
-			return nil, err
+
+		// if there are no keys to retry, then we're done.
+		if len(retryKeys) == 0 {
+			return false, nil, resultAccumulator
 		}
 
-		for _, item := range items {
-			getOp, ok := item.(*gocb.GetOp)
-			if !ok {
-				continue
+		// otherwise, retry the keys the need to be retried
+		keys = retryKeys
+
+		// return true to signal that this function needs to be retried
+		return true, nil, nil
+
+	}
+
+	return worker
+
+}
+
+func (bucket CouchbaseBucketGoCB) processBatch(keys []string, resultAccumulator map[string][]byte, retryKeys []string) error {
+
+	var items []gocb.BulkOp
+	for _, key := range keys {
+		var value []byte
+		item := &gocb.GetOp{Key: key, Value: &value}
+		items = append(items, item)
+	}
+	err := bucket.Do(items)
+	if err != nil {
+		return err
+	}
+
+	for _, item := range items {
+		getOp, ok := item.(*gocb.GetOp)
+		if !ok {
+			continue
+		}
+		// Ignore any ops with errors.
+		// NOTE: some of the errors are misleading:
+		// https://issues.couchbase.com/browse/GOCBC-64
+		if getOp.Err == nil {
+			byteValue, ok := getOp.Value.(*[]byte)
+			if ok {
+				resultAccumulator[getOp.Key] = *byteValue
 			}
-			// Ignore any ops with errors.
-			// NOTE: some of the errors are misleading:
-			// https://issues.couchbase.com/browse/GOCBC-64
-			if getOp.Err == nil {
-				byteValue, ok := getOp.Value.(*[]byte)
-				if ok {
-					result[getOp.Key] = *byteValue
-				}
+		} else {
+			// if it's a recoverable error, then throw it in retry collection.
+			if isRecoverableGoCBError(getOp.Err) {
+				retryKeys = append(retryKeys, getOp.Key)
 			}
 		}
+
+	}
+
+	return nil
+}
+
+func createBatchesKeys(batchSize int, keys []string) [][]string {
+
+	// boundary checking
+	if len(keys) == 0 {
+		Warn("createBatchesKeys called with empty keys")
+		return [][]string{}
+	}
+	if batchSize == 0 {
+		Warn("createBatchesKeys called with invalid batchSize")
+		result := [][]string{}
+		return append(result, keys)
+	}
+
+	numBatches := len(keys) / batchSize
+	if len(keys)%batchSize > 0 {
+		// need another batch for the residue
+		numBatches += 1
+	}
+	batches := make([][]string, numBatches)
+
+	for idxBatch, batch := range batches {
+		for _, key := range keys {
+			batch = append(batch, key)
+			keys = keys[1:]
+			if len(batch) == batchSize {
+				// this batch is full
+				break
+			}
+		}
+		batches[idxBatch] = batch
+	}
+
+	return batches
+
+}
+
+func isNotFoundError(err error) bool {
+	if err == nil {
+		// not even an error!
+		return false
+	}
+	return strings.Contains(err.Error(), "Key not found")
+}
+
+func shouldRetryGoCBOp(err error) bool {
+
+	if err == nil {
+		// no error, no need to retry
+		return false
+	}
+
+	return isRecoverableGoCBError(err)
+
+}
```

Comment:
Is there value in having separate methods for shouldRetryGoCBOp and isRecoverableGoCBError?